### PR TITLE
feat: support json inside temporal values

### PR DESCRIPTION
**Pull request by thomasBousselin**

### CODE REVIEW

File: search-service/src/test/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceServiceTests.kt [hunk 1]
```diff
@@ -609,7 +611,7 @@ class AttributeInstanceServiceTests : WithTimescaleContainer, WithKafkaContainer
             assertThat(results)
                 .singleElement()
                 .matches {
-                    (it as SimplifiedAttributeInstanceResult).value == 100.0
+                    (it as SimplifiedAttributeInstanceResult).value == 100.1
```
Comment (thomasBousselin) [inline search-service/src/test/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceServiceTests.kt:618]: This is a bit cheating but the `.0` information is lost when the data is stored in measured_value. Basically the old code always return the `.0`.

I posted :
```json
{
  "id": "urn:11",
  "type": "Test",
  "property":{
    "type": "Property",
    "value": 20.0
  },
    "property2":{
    "type": "Property",
    "value": 20
  }
}
```

And a get with timeproperty=createdAt&format=temporalValues
returned :
```json
{
  "id": "urn:11",
  "type": "Test",
  "createdAt": "2026-04-02T14:19:43.406048Z",
  "property2": {
    "type": "Property",
    "values": [
      [
        20.0,
        "2026-04-02T14:19:43.406048Z"
      ]
    ]
  },
  "property": {
    "type": "Property",
    "values": [
      [
        20.0,
        "2026-04-02T14:19:43.406048Z"
      ]
    ]
  }
}
```
So this small edge case mistake just changed to another one.

This might be fixed by only relying on `value` and not on `measured_value` anymore. I will test this on a specific PR.

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+ALTER TABLE attribute_instance
+    ALTER COLUMN value TYPE JSONB
+        USING value::jsonb;
+
+ALTER TABLE attribute_instance_audit
+    ALTER COLUMN value TYPE JSONB
+        USING value::jsonb;
+
+UPDATE attribute_instance
```
Comment (bobeal) [inline search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: Missing update for `attribute_instance_audit`

File: search-service/src/test/kotlin/com/egm/stellio/search/common/DBConversionUtilsTests.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.egm.stellio.search.common
+
+import com.egm.stellio.search.common.util.toJson
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Test
+import org.springframework.test.context.ActiveProfiles
+import kotlin.test.assertEquals
+
+@ActiveProfiles("test")
+class DBConversionUtilsTests {
```
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/common/DBConversionUtilsTests.kt:11]: may worth adding tests for other types?

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql (hunk 2 above)
Comment (bobeal) [reply search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: Also, it will probably take some time on large DBs. A warning to add in the upgrade doc?

File: search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt [hunk 4]
```diff
@@ -22,14 +22,15 @@ fun toOptionalZonedDateTime(entry: Any?): ZonedDateTime? =
...
 
 fun Json.deserializeExpandedPayload(): Map<String, List<Any>> = this.asString().deserializeExpandedPayload()
 fun Json.deserializeAsMap(): Map<String, Any> = this.asString().deserializeAsMap()
+fun Json.deserializeTemporalValue(): Any = deserializeAs<Any>(this.asString())
 
-fun ExpandedAttributeInstance.toJson(): Json = Json.of(serializeObject(this))
+fun Any.toJson(): Json = Json.of(serializeObject(this))
 
 fun valueToDoubleOrNull(value: Any): Double? =
     when (value) {
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt:37]: `valueToStringOrNull` below is no longer used

File: search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt [hunk 5]
```diff
@@ -22,14 +22,15 @@ fun toOptionalZonedDateTime(entry: Any?): ZonedDateTime? =
     (entry as? OffsetDateTime)?.atZoneSameInstant(ZoneOffset.UTC)
 fun <T> toList(entry: Any?): List<T> = (entry as Array<T>).toList()
 fun <T> toOptionalList(entry: Any?): List<T>? = (entry as? Array<T>)?.toList()
-fun toJson(entry: Any?): Json = entry as Json
+fun castToJson(entry: Any?): Json = entry as Json
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt]: this kind of breaks the naming conventions for all these methods

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/util/AttributeUtils.kt [hunk 6]
```diff
@@ -60,11 +61,12 @@ fun NgsiLdAttributeInstance.toAttributeMetadata(): Either<APIException, Attribut
             guessPropertyValueType(this).let {
                 Triple(AttributeType.Property, it.first, it.second)
             }
+
         is NgsiLdRelationshipInstance ->
             Triple(
                 AttributeType.Relationship,
                 Attribute.AttributeValueType.URI,
-                Triple(this.objectId.toString(), null, null)
+                Triple(this.objectId.toString().toJson(), null, null)
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/entity/util/AttributeUtils.kt]: you need `.toString().toJson()` or `.toJson()` could be enough?

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/util/AttributeUtils.kt [hunk 7]
```diff
@@ -124,23 +126,25 @@ fun guessAttributeValueType(
...
-        is String -> Pair(Attribute.AttributeValueType.STRING, Triple(value, null, null))
-        is Boolean -> Pair(Attribute.AttributeValueType.BOOLEAN, Triple(value.toString(), null, null))
-        is LocalDate -> Pair(Attribute.AttributeValueType.DATE, Triple(value.toString(), null, null))
-        is ZonedDateTime -> Pair(Attribute.AttributeValueType.DATETIME, Triple(value.toString(), null, null))
-        is LocalTime -> Pair(Attribute.AttributeValueType.TIME, Triple(value.toString(), null, null))
-        else -> Pair(Attribute.AttributeValueType.STRING, Triple(value.toString(), null, null))
+        is Map<*, *> ->
+            Pair(Attribute.AttributeValueType.OBJECT, Triple(value.toJson(), null, null))
+        is List<*> ->
+            Pair(Attribute.AttributeValueType.ARRAY, Triple(value.toJson(), null, null))
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/entity/util/AttributeUtils.kt]: Unnecessary CRs. (make the reading less easy)

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt [hunk 8]
```diff
@@ -272,8 +276,9 @@ class AttributeInstanceService(
         val valueColumn = when {
             // for deletedAt, the NGSI-LD Null representation is always stored as string in value column
             temporalQuery.timeproperty == DELETED_AT -> "value"
-            attributes[0].attributeValueType == AttributeValueType.NUMBER -> "measured_value as value"
-            attributes[0].attributeValueType == AttributeValueType.GEOMETRY -> "public.ST_AsText(geo_value) as value"
+            attributes[0].attributeValueType == AttributeValueType.NUMBER -> "to_jsonb(measured_value) as value"
+            attributes[0].attributeValueType == AttributeValueType.GEOMETRY ->
+                "to_jsonb(geo_value) as value"
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt]: Unnecessary CRs. (make the reading less easy)

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt [hunk 9]
```diff
@@ -338,13 +343,16 @@ class AttributeInstanceService(
...
-                // in this case, just display an empty value (something happened, but we can't display it)
-                value = row["value"] ?: "",
-                time = toZonedDateTime(row["start"])
-            )
+            TemporalRepresentation.TEMPORAL_VALUES -> {
+                val value = row["value"]
+                SimplifiedAttributeInstanceResult(
+                    attributeUuid = toUuid(row["temporal_entity_attribute"]),
+                    // the type of the value of a property may have changed in the history (e.g., from number to string)
+                    // in this case, just display an empty value (something happened, but we can't display it)
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt]: to be updated

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 8 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt]: `to_json` on a `geometry` value works well?

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/TemporalQueryService.kt [hunk 10]
```diff
@@ -199,20 +197,6 @@ class TemporalQueryService(
             }.mapValues {
                 attributeInstanceService.search(temporalEntitiesQuery, it.value, origin).bind()
             }
-            .mapValues {
-                // when retrieved from DB, values of geo-properties are encoded in WKT and won't be automatically
-                // transformed during compaction as it is not done for temporal values, so it is done now
-                if (it.key == Attribute.AttributeValueType.GEOMETRY &&
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/TemporalQueryService.kt:205]: where has this moved to?

File: search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt (hunk 5 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt]: Yes, but Any.toJson() and toJson(Any) have the same JVM signature. So i must rename one of them. If you see a better naming let me know.

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql (hunk 2 above)
Comment (thomasBousselin) [reply search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: Hum maybe, let's see how much time it take on dev.

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 8 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt]: Yes, it returns the geojson. (cf [test with geometry](https://github.com/stellio-hub/stellio-context-broker/pull/1690/changes/BASE..d92faab4fe23fc2707a20f69ec80648c91159071#diff-8cbf1f81db9609be9d09b940377429e5d4df520c77e681c1ff2d88ed675fa089R1256))

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/TemporalQueryService.kt (hunk 10 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/TemporalQueryService.kt:205]: This it taken care of by the `to_jsonb(geo_value)` in composeSearchSelectStatement.
See [this test](https://github.com/stellio-hub/stellio-context-broker/pull/1690/changes/BASE..d92faab4fe23fc2707a20f69ec80648c91159071#diff-8cbf1f81db9609be9d09b940377429e5d4df520c77e681c1ff2d88ed675fa089R1256)

File: search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt (hunk 5 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt]: I suggest renaming `Any.toJson` to `Any.asJsonObject()`

**Comment by bobeal** [comment]
There are failing tests in the NGSI-LD Test Suite.

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql [hunk 11]
```diff
@@ -0,0 +1,23 @@
+ALTER TABLE attribute_instance
+    ALTER COLUMN value TYPE JSONB
+        USING value::jsonb;
```
Comment (bobeal) [inline search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: This is going to fail on existing data.

Just did a quick test on a copy of an `attribute_instance_audit` table:

```
stellio_search=# ALTER TABLE attribute_instance_audit_test
ALTER COLUMN value TYPE JSONB USING value::jsonb;
ERROR:  invalid input syntax for type json
DETAIL:  Token "varValue" is invalid.
CONTEXT:  JSON data, line 1: varValue
```

Use `to_jsonb` instead?

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    'https://uri.etsi.org/ngsi-ld/VocabProperty'
+);
+
+UPDATE attribute_instance_audit
+SET value = (value::text::jsonb #>> '{}')::jsonb
+WHERE payload #>> '{@type, 0}' in (
+                                   'https://uri.etsi.org/ngsi-ld/LanguageProperty',
+                                   'https://uri.etsi.org/ngsi-ld/JsonProperty',
+                                   'https://uri.etsi.org/ngsi-ld/VocabProperty'
+    );
```
Comment (bobeal) [inline search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: Wild indentation

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt [hunk 13]
```diff
@@ -272,8 +276,8 @@ class AttributeInstanceService(
         val valueColumn = when {
             // for deletedAt, the NGSI-LD Null representation is always stored as string in value column
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt]:
```suggestion
            // for deletedAt, the NGSI-LD Null representation is always stored as jsonb in value column
```

File: search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt [hunk 14]
```diff
@@ -28,19 +28,13 @@ fun toInt(entry: Any?): Int = (entry as Long).toInt()
 
 fun Json.deserializeExpandedPayload(): Map<String, List<Any>> = this.asString().deserializeExpandedPayload()
 fun Json.deserializeAsMap(): Map<String, Any> = this.asString().deserializeAsMap()
+fun Json.deserializeTemporalValue(): Any = deserializeAs<Any>(this.asString())
 
-fun ExpandedAttributeInstance.toJson(): Json = Json.of(serializeObject(this))
+fun Any.asJsonObject(): Json = Json.of(serializeObject(this))
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/common/util/DBConversionUtils.kt]: Using "JSON object" can be confusing (it is for me, because I see ... a JSON object). `asJsonb`? `asJsonbValue`?

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt [hunk 15]
```diff
@@ -272,8 +276,8 @@ class AttributeInstanceService(
...
             temporalQuery.timeproperty == DELETED_AT -> "value"
-            attributes[0].attributeValueType == AttributeValueType.NUMBER -> "measured_value as value"
-            attributes[0].attributeValueType == AttributeValueType.GEOMETRY -> "public.ST_AsText(geo_value) as value"
+            attributes[0].attributeValueType == AttributeValueType.NUMBER -> "to_jsonb(measured_value) as value"
+            attributes[0].attributeValueType == AttributeValueType.GEOMETRY -> "to_jsonb(geo_value) as value"
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:281]: `to_jsonb` does the conversion from WKB to GeoJson, mind blowing!

there is still the conversion when asking for the normalized representation, still needed?

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt [hunk 16]
```diff
@@ -338,13 +342,14 @@ class AttributeInstanceService(
...
-                // the type of the value of a property may have changed in the history (e.g., from number to string)
-                // in this case, just display an empty value (something happened, but we can't display it)
-                value = row["value"] ?: "",
-                time = toZonedDateTime(row["start"])
-            )
+            TemporalRepresentation.TEMPORAL_VALUES -> {
+                val value = row["value"]
+                SimplifiedAttributeInstanceResult(
+                    attributeUuid = toUuid(row["temporal_entity_attribute"]),
+                    value = toJson(value).deserializeTemporalValue(),
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:350]: no problem if `value` is null?

**Review by bobeal — COMMENTED** [review]
There are 3 failing TCs in the NGSI-LD Test Suite. Because `TemporalEntityBuilder` tries to cast to string values that are a list or a map (for `LanguageProperty`, `JsonProperty` and `VocabProperty`).

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 16 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:350]: It will crash, but value is not supposed to be null because both 'measured_value' and 'geo-value' are renamed 'value' in the request.
```
      attributes[0].attributeValueType == AttributeValueType.NUMBER -> "to_jsonb(measured_value) as value"
      attributes[0].attributeValueType == AttributeValueType.GEOMETRY -> "to_jsonb(geo_value) as value"
```

File: search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql (hunk 11 above)
Comment (thomasBousselin) [reply search-service/src/main/resources/db/migration/V0_52__temporal_value_as_jsonb.sql]: fixed

**Comment by thomasBousselin** [comment]
> There are 3 failing TCs in the NGSI-LD Test Suite. Because TemporalEntityBuilder tries to cast to string values that are a list or a map (for LanguageProperty, JsonProperty and VocabProperty).

fixed

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 15 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:281]: ping

File: search-service/src/main/resources/db/migration/V0_54__temporal_value_as_jsonb.sql [hunk 17]
```diff
@@ -0,0 +1,27 @@
+ALTER TABLE attribute_instance
+    ALTER COLUMN value TYPE JSONB
+        USING to_jsonb(value);
+
+ALTER TABLE attribute_instance_audit
+    ALTER COLUMN value TYPE JSONB
+        USING to_jsonb(value::jsonb);
```
Comment (bobeal) [inline search-service/src/main/resources/db/migration/V0_54__temporal_value_as_jsonb.sql]: It should be `to_jsonb(value)` (`value::jsonb` is going to fail)

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 15 above)
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:281]: using to_jsonb here was the perfect scenario since I had to receive a json anyway and the code to tranform to json was specific for the temporal case. It could be interesting to use this in other places but not as much as here.

File: search-service/src/main/resources/db/migration/V0_54__temporal_value_as_jsonb.sql (hunk 17 above)
Comment (thomasBousselin) [reply search-service/src/main/resources/db/migration/V0_54__temporal_value_as_jsonb.sql]: forgot this one.

File: search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt (hunk 15 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/temporal/service/AttributeInstanceService.kt:281]: that does not answer my question, but I'll take it as a yes...

**Comment by bobeal** [comment]
why is Anchore unhappy?

**Review by bobeal — DISMISSED** [review]

**Comment by thomasBousselin** [comment]
No changes, just squashed the commits to keep the informations from the MR.

**Review by bobeal — APPROVED** [review]
And take care of doing a "Rebase and merge" and not a "Squash and merge"
